### DISCUSSION TITLE
STL Boundary faces

### DISCUSSION THREAD

**Discussion by f-pucci** [pyvista/pyvista#4727] [Q&A]
Hello everyone.

I have developed a .stl geometry using pyvista to be used for CFD simulations in OpenFOAM.

I need to define the so-called 'boundary patches' to identify the inlet, outlet and maybe wall faces. 
Is there a way to do it with pyvista?
Maybe a bounding box selection kind of thing and save the data as an 'integrated stl' (?).

The idea would be using just python modules or simple bash scripts (to run OpenFOAM).

Thank you in advance for your help!

**Comment by edabor** [comment]
Hello @f-pucci,

If you are allowed to do so, could you provide us the `.stl` file so that we can have a look ?

In the meantime, you can try to play with the combination of `compute_normals` and `connectivity` methods.
See the example below.

The most important parameter is the `feature_angle`, which helps the `connectivity` filter to compute region ids that you can latter attach boundary conditions to each value.
You need to iteratively find the `feature_angle` value that produces the desired regions.
Please note that with `split_vertices=True`, points are duplicated at the regions borders.

```python
from pyvista.examples import load_nut

c = load_nut().compute_normals(feature_angle=60, split_vertices=True).connectivity()
c.plot()
```
Here are the documentation links:
https://docs.pyvista.org/version/dev/api/core/_autosummary/pyvista.PolyDataFilters.compute_normals.html#pyvista.PolyDataFilters.compute_normals
https://docs.pyvista.org/version/dev/api/core/_autosummary/pyvista.DataSetFilters.connectivity.html

PS: If your 3D model is simple enough, I would recommend you to use a dedicated CAD software, since they allow a finer control on how to specify boundary conditions to specific regions. gmsh is one of them, as I already recommended you [here](https://github.com/pyvista/pyvista/discussions/4616#discussioncomment-6431269)

**Reply by f-pucci** [reply]
Hello, 

thank you for your answers!
Unfortunately I cannot really share the .stl file. 

The geometry is a water-tight manifold geometry made of a single component.
I generated it by an implicit mathematical model. Then, I have extracted the upper and lower faces, which I used for an extrusion and obtain a fluid domain like:
![image](https://github.com/pyvista/pyvista/assets/126581990/15f17c35-4391-4a28-86a9-fea2a0ade3e9)


I have tried the compute_normals/connectivity approach, but it doesn't really 'split' the geometry. If I plot it, everything has the same 'colour'. I tried with feature angles = 40, 50 , 60, 70, 80, 89 and 90.

Any idea?

**Reply by MatthewFlamm** [reply]
How do you get the stl file to begin with?  Can you get it to provide separate stl files for each face?  I suspect that any method in PyVista will be brittle to small differences in the input geometry.

**Reply by f-pucci** [reply]
I generate an unstructured grid, then with the data clip function I extract the geometry based on the isovalue. 
Then I fix the geometry with pymeshfix and remesh it with another module. Then I extrude the top and bottom face and remesh these extrusions too.

**Reply by MatthewFlamm** [reply]
If the top and bottom face are the boundary faces of interest, you might already have them.

**Reply by f-pucci** [reply]
I have managed to extract the outer faces of the extrusions by doing, for example:

upperPoints= np.arange(mesh.n_points)[mesh.points[:,2]==upperZBound]
lowerPoints= np.arange(mesh.n_points)[mesh.points[:,2]==lowerZBound]

removePoints = np.concatenate((lowerPoints, upperPoints))

lowerCells = mesh.extract_points(lowerPoints, adjacent_cells=False)

wallCells = mesh.remove_points(removePoints, mode='all')[0]

So that I can have the two extreme faces and then the rest of the geometry.

Do you know of a way to remove those floating elements?
![image](https://github.com/pyvista/pyvista/assets/126581990/7233a073-6160-43bb-9e96-43f338f810f6)

If I do extract_largest(), i can manage to remove some of them, but not all. And if I do connectivity, I can see that some are connected to the body.
Those elements probably come from the boolean union with a cylinder.
